tests/unit/Fun/LazyTest.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Tests\Unit\Fun;
+
+use Generator;
+use PHPUnit\Framework\TestCase;
+use Psl\Fun;
+use RuntimeException;
+
+use function pow;
+
+final class LazyTest extends TestCase
+{
+    public function testItDoesNotLoadInitializerDuringCreation(): void
+    {
+        $this->expectNotToPerformAssertions();
+        Fun\lazy(static function () {
+            throw new RuntimeException('nonono');
+        });
+    }
+
+    public function testItCanBeUsedAsALazyProxy(): void
+    {
+        $proxy = Fun\lazy(static fn (): int => 132);
+
+        static::assertSame(132, $proxy());
+    }
+
+    public function testItCanBeUsedAsALazyProxyAndChainCalls(): void
+    {
+        $x = new class {
+            public function doSomething(): int
+            {
+                return 132;
+            }
+        };
+        $proxy = Fun\lazy(static fn (): object => $x);
+
+        static::assertSame($x, $proxy());
+        static::assertSame($x, $proxy());
+
+        static::assertSame(132, $proxy()->doSomething());
+        static::assertSame(132, $proxy()->doSomething());
+    }
+
+    public function testItCanDealWithNullValues(): void
+    {
+        $counter = (object)(['tick' => 0]);
```

Comment:
huh, we should probably add `Psl\Ref` 😛 https://docs.hhvm.com/hsl/reference/class/HH.Lib.Ref/

Reply:
It is not completely clear from that page how to use it.
Maybe rather something like react's `useRef()` ? 

https://reactjs.org/docs/hooks-reference.html#useref

```php
$pointer = useRef(0);
$pointer->current(); // 0
$pointer->current(fn ($previous) => 1); // 1
$pointer->current(); // 1
```

Reply:
it's basically:

```php
/**
 * @template T
 */
final class Ref {
  /**
   * @param T $value
   */
  public function __construct(
    public mixed $value
  ) {}
}
```

Reply:
https://github.com/hhvm/hsl/blob/v3.28.x/src/Ref.php#L25

Reply:
Very similar to react's useRef.
Sounds like something that could be useful indeed!